br/pkg/lightning/backend/local/local.go

```diff
@@ -1541,53 +1604,89 @@ func (local *local) writeAndIngestByRanges(ctx context.Context, engine *Engine,
 	var allErr error
 	var wg sync.WaitGroup
 	metErr := atomic.NewBool(false)
+	storeId2Range := make(map[uint64][]Range)
+	storeId2RangeIndex := make(map[uint64]int)
 
 	for _, r := range ranges {
 		startKey := r.start
 		endKey := r.end
-		w := local.rangeConcurrency.Apply()
-		// if meet error here, skip try more here to allow fail fast.
-		if metErr.Load() {
-			local.rangeConcurrency.Recycle(w)
-			break
+		regions, err := local.scanRegion(ctx, engine, startKey, endKey, regionSplitSize, regionSplitKeys)
+		if err != nil {
+			log.FromContext(ctx).Warn("scaen regino failed", zap.Error(err))
+			return err
 		}
-		wg.Add(1)
-		go func(w *worker.Worker) {
-			defer func() {
-				local.rangeConcurrency.Recycle(w)
-				wg.Done()
-			}()
-			var err error
-			// max retry backoff time: 2+4+8+16+30*26=810s
-			backOffTime := time.Second
-			for i := 0; i < maxWriteAndIngestRetryTimes; i++ {
-				err = local.writeAndIngestByRange(ctx, engine, startKey, endKey, regionSplitSize, regionSplitKeys)
-				if err == nil || common.IsContextCanceledError(err) {
-					return
-				}
-				if !local.isRetryableImportTiKVError(err) {
-					break
-				}
-				log.FromContext(ctx).Warn("write and ingest by range failed",
-					zap.Int("retry time", i+1), log.ShortError(err))
-				backOffTime *= 2
-				if backOffTime > maxRetryBackoffTime {
-					backOffTime = maxRetryBackoffTime
-				}
-				select {
-				case <-time.After(backOffTime):
-				case <-ctx.Done():
-					return
-				}
+		// only take to leader's store id fo the first region
+		if len(regions) > 0 {
```

Comment:
one key range may have multiple regions, because regions are split by other logic previously. That's why writeAndIngestByRange we have a separate PaginateScanRegion and handle each region in writeAndIngestPairs.

So your current PR is not very precise. It's still an improvement.

Reply:
yeah, I just want to make the logic simple so I only consider the leader store id of the first region. I'm trying a solution which consider all peer id of all regions. I'm currently testing its performance

Reply:
updated in : https://github.com/pingcap/tidb/pull/54976